samples/connectedhomeip/lock/CMakeLists.txt

```diff
@@ -0,0 +1,45 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+cmake_minimum_required(VERSION 3.13.1)
+
+get_filename_component(CHIP_ROOT ../../../../modules/lib/connectedhomeip REALPATH)
+get_filename_component(COMMON_ROOT ../common REALPATH)
+
+include(${CHIP_ROOT}/config/nrfconnect/app/app.cmake)
+
+project(chip-lock)
+
+target_include_directories(app PRIVATE src ${COMMON_ROOT}/src)
+
+target_sources(app PRIVATE
+               src/app_task.cpp
+               src/bolt_lock_manager.cpp
+               src/gen/call-command-handler.cpp
+               src/gen/callback-stub.cpp
+               src/main.cpp
+               src/zcl_callbacks.cpp
+               ${COMMON_ROOT}/src/led_widget.cpp
+               ${COMMON_ROOT}/src/nfc_widget.cpp
+               ${COMMON_ROOT}/src/thread_util.cpp
+               ${CHIP_ROOT}/src/app/reporting/reporting-default-configuration.cpp
+               ${CHIP_ROOT}/src/app/reporting/reporting.cpp
+               ${CHIP_ROOT}/src/app/server/DataModelHandler.cpp
+               ${CHIP_ROOT}/src/app/util/af-event.cpp
+               ${CHIP_ROOT}/src/app/util/af-main-common.cpp
+               ${CHIP_ROOT}/src/app/util/attribute-size.cpp
+               ${CHIP_ROOT}/src/app/util/attribute-storage.cpp
+               ${CHIP_ROOT}/src/app/util/attribute-table.cpp
+               ${CHIP_ROOT}/src/app/util/binding-table.cpp
+               ${CHIP_ROOT}/src/app/util/chip-message-send.cpp
+               ${CHIP_ROOT}/src/app/util/client-api.cpp
+               ${CHIP_ROOT}/src/app/util/ember-print.cpp
+               ${CHIP_ROOT}/src/app/util/message.cpp
+               ${CHIP_ROOT}/src/app/util/process-cluster-message.cpp
+               ${CHIP_ROOT}/src/app/util/process-global-message.cpp
+               ${CHIP_ROOT}/src/app/util/util.cpp
+               ${CHIP_ROOT}/src/app/clusters/bindings/bindings.cpp
+               ${CHIP_ROOT}/src/app/clusters/on-off-server/on-off.cpp)
```

Comment:
@carlescufi @mbolivar-nordic should we consider to fork: https://github.com/project-chip/connectedhomeip for the sole purpose of adding: `zephyr/module.yml` with content:
```
build:
  cmake-ext: True
  kconfig-ext: True
```
that would allow a more seamless inclusion into NCS, and at the same time be able to use `ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR` instead of: `../../../../modules/lib/connectedhomeip` (the latter effectively moves to `west` topdir, and thus is very dependent on the structure of the `west` workspace instead using the Zephyr modules var.)

Reply:
@tejlmand `app.cmake` file, which CHIP samples in NCS include, adds CHIP as an out-of-tree Zephyr module. If I understand correctly, by following your solution we would end up with CHIP being included once as a regular module (but with CMake and Kconfig not loaded) and then as an out-of-tree module. Can there be any problems with such a design?

Note that we've decided with @carlescufi to include `app.cmake` file directly in the samples instead of adding west manifest to https://github.com/nrfconnect/sdk-connectedhomeip fork because we would like to keep the possibility to build nRF Connect samples from the CHIP upstream, even when using nRF Connect SDK version which already contains CHIP as a subproject. In other words, nRF Connect samples from the CHIP repository should use CHIP libraries built using the same CHIP revision. As far as I know, Zephyr currently doesn't support configuring modules on a project-basis, so there's no way to blacklist the NCS-supplied CHIP in case a project needs to use a different CHIP version, right?

Reply:
> @tejlmand app.cmake file, which CHIP samples in NCS include, adds CHIP as an out-of-tree Zephyr module. 

Right, I didn't look into this file: https://github.com/nrfconnect/sdk-connectedhomeip/blob/master/config/nrfconnect/app/app.cmake

Actually, this file mixes things, and as minimum it should only do:
```
list(APPEND ZEPHYR_EXTRA_MODULES ${CMAKE_CURRENT_LIST_DIR}/../chip-module)
```
That way, the chip module would still be a Zephyr module, and all other code will follow existing pattern of
```
find_package(Zephyr)
```
and you should be able to use the `APP_LINK_WITH_CHIP` Kconfig symbol if done properly, instead of:
```
target_link_libraries(app PUBLIC chip)
```

> As far as I know, Zephyr currently doesn't support configuring modules on a project-basis, so there's no way to blacklist the NCS-supplied CHIP in case a project needs to use a different CHIP version, right?

Even though this was never considered a use-case, then if we are using the new Zephyr module ext root, toghether with the name key, then `ZEPHYR_EXTRA_MODULES` should actually overrule the module from the manifest, which is what I understand you want to do (using the locally cloned version instead of NCS manifest specified), and still building the project using rest of NCS.

Let me do some initial testing and get back to this, as I believe this might be possible with minimal effort.

Reply:
@Damian-Nordic I have some initial cleanup ready regarding CMake, but i'm having build problems when checking out this PR.

```
[246/260] Performing configure step for 'chip-gn'
ERROR at //third_party/connectedhomeip/third_party/pigweed/repo/pw_build/target_types.gni:46:3: Source file not found.
  source_set(target_name) {
  ^------------------------
The target:
  //third_party/connectedhomeip/third_party/pigweed/repo/third_party/nanopb:nanopb
has a source file:
  //third_party/connectedhomeip/third_party/nanopb/repo/pb.h
which was not found.
```
seems like something is missing between `//` first in path.

I also noticed this:
https://github.com/nrfconnect/sdk-connectedhomeip/blob/b430c424ab62438fffae71c9bd2f23d3f0d442d2/config/nrfconnect/chip-module/CMakeLists.txt#L212
```
    CONFIGURE_COMMAND       gn --root=${CHIP_ROOT}/config/nrfconnect/chip-gn gen --check --fail-on-unused-args ${CMAKE_CURRENT_BINARY_DIR}
```
Please use `find_program()` in CMake to look up required programs, so that users are informed of a missing program **before** compilation.

When building openthread based samples, it fails with:
```
../../../../../../config/nrfconnect/chip-gn/third_party/connectedhomeip/src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp: In member function 'CHIP_ERROR chip::DeviceLayer::Internal::GenericThreadStackManagerImpl_OpenThread< <template-parameter-1-1> >::DoInit(otInstance*)':
../../../../../../config/nrfconnect/chip-gn/third_party/connectedhomeip/src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:845:18: error: 'otLinkModeConfig' {aka 'struct otLinkModeConfig'} has no member named 'mSecureDataRequests'
  845 |         linkMode.mSecureDataRequests = true;
      |                  ^~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
[105/169] c++ obj/third_party/connectedhomeip/src/platform/libDeviceLayer.SystemTimerSupport.cpp.o

```

On top of this, there are a couple of issues related to Zephyr build integration that I have fixed up.

Reply:
Have created an example on how to do this in a more proper way:
https://github.com/tejlmand/fw-nrfconnect-nrf/tree/review/3670_proposal
https://github.com/tejlmand/sdk-connectedhomeip/tree/review/3670_proposal
https://github.com/tejlmand/fw-nrfconnect-zephyr-1/tree/review/3670_proposal

Some explanation:
As you can see:
https://github.com/tejlmand/fw-nrfconnect-nrf/commit/cce4663687cfe364272f5e3d0bc2b649ff18941c
no longer needs to know the structure of the `west` workspace.
It uses `find_package(Zephyr)` like any other sample, and relies on `ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR` instead.
It doesn't need the `app.cmake` file anymore.

In https://github.com/tejlmand/sdk-connectedhomeip/tree/review/3670_proposal there is a series of commits, but the important part here is now this commit:
https://github.com/tejlmand/sdk-connectedhomeip/commit/814094ab97b53cdf295eee580b550ac6b68f2e29

as this commit, together with this:
https://github.com/tejlmand/fw-nrfconnect-zephyr-1/commit/e4d3b94edd8744ab4c6e3d3fd93b0174bdf50b8f

will allow the connected home ip repo to decide the exact repo to use when building a sample in this repo, as you can see if you print the module dir:
Example
```
$ pwd
/projects/github/ncs/modules/lib/connectedhomeip
$ cd examples/lock-app/nrfconnect/
$ cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -Bbuild . 
Including boilerplate (Zephyr workspace): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/ncs/modules/lib/connectedhomeip/examples/lock-app/nrfconnect
...
-- ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR: /projects/github/ncs/modules/lib/connectedhomeip/config/nrfconnect/app/../../..
-- Configuring done

# Copy the repo to custom:
$ cd -
$ cp -a ../connectedhomeip/ ../../../custom_connectedhomeip/
$ cd ../../../custom_connectedhomeip/examples/lock-app/nrfconnect/
$ cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -Bbuild . 
Including boilerplate (Zephyr workspace): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/ncs/custom_connectedhomeip/examples/lock-app/nrfconnect
...
-- ZEPHYR_CONNECTEDHOMEIP_MODULE_DIR: /projects/github/ncs/custom_connectedhomeip/config/nrfconnect/app/../../..
```
so that when building any sample inside the repo, that repo is used as the Zephyr module, and any sample outside the repo will use the repo in the manifest, unless overruled be `-DZEPHYR_EXTRA_MODULES=<module>`.

With those patches, there is basically no longer a need for the `app.cmake`, as I think those lines should be removed:
https://github.com/nrfconnect/sdk-connectedhomeip/blob/master/config/nrfconnect/app/app.cmake#L25-L27
```
if (EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/dts.overlay")
    set(DTC_OVERLAY_FILE "${CMAKE_CURRENT_SOURCE_DIR}/dts.overlay")
endif()
```
as this is not the best way to do this, but I believe that should be reserved for a separate discussion.

I also noticed several other parts in https://github.com/nrfconnect/sdk-connectedhomeip that will benefit from a cleanup, but I guess that is outside this discussion.

Reply:
@tejlmand Thanks for the suggestion. Some problems you've mentioned have already been fixed on the CHIP upstream and we need to update the fork to pull them into NCS.

Overall I like the proposal, but if I understand correctly, if we follow your advice we first need to contribute your changes to Zephyr before we can move on with the integration of CHIP. I'm a bit concerned if we manage to do so before the code freeze. Or maybe you propose to push your Zephyr changes to the sdk-zephyr fork only for now?